### ISSUE TITLE
User Story

### ISSUE THREAD

**Issue by ajsarkar28** [department-of-veterans-affairs/va-mobile-app#6816]
## Description 
As a mobile team member, I want to understand how long it takes for UI to render on the home screen, so that I have a better sense of how adding more data to the home screen will affect performance.

## Acceptance Criteria
-  [ ] A mechanism is developed which can record the time it takes for the front end to load on the home screen
-  [ ] Home screen load times are recorded

**Comment by Sparowhawk** [comment]
Since this looks like we are trying to determine a load time and this might be something we need to maintain. I would suggest whoever picks this up create something in demo mode where we create a new screen APILoadTimeScreen. Where we load everything we possibly can load and display the load time for each one and display the requested data for mean, median, low, and high times.

**Comment by alexandec** [comment]
## Methodology

I looked through all the GET requests in the app, compiled a list of all the relevant ones, then added the following code to the HomeScreen component to benchmark loading a variety of endpoints at once:

```
const timeRequest = async (endpoint: string) => {
  const start = performance.now()
  await get(endpoint)
  const end = performance.now()
  console.log(`${endpoint}: ${Math.round(end - start)}ms`)
}

useEffect(() => {
  const startAll = performance.now()
  Promise.all([
    timeRequest('/v0/appointments'),
    timeRequest('/v0/claims-and-appeals-overview'),
    timeRequest('/v0/claims/decision-letters'),
    timeRequest('/v0/disability-rating'),
    timeRequest('/v0/facilities-info'),
    timeRequest('/v0/health/rx/prescriptions'),
    timeRequest('/v0/letters'),
    timeRequest('/v0/letters/beneficiary')
    timeRequest('/v0/maintenance_windows'),
    timeRequest('/v0/messaging/health/folders'),
    timeRequest('/v0/messaging/health/folders/0/messages'),
    timeRequest('/v0/messaging/health/recipients'),
    timeRequest('/v0/military-service-history'),
    timeRequest('/v0/payment-history'),
    timeRequest('/v0/payment-information/benefits'),
    timeRequest('/v0/user/authorized-services'),
    timeRequest('/v0/user/demographics'),
    timeRequest('/v0/user/gender_identity/edit'),
    timeRequest('/v1/health/immunizations'),
    timeRequest('/v1/user'),
  ]).then(() => {
    const endAll = performance.now()
    console.log(`All requests finished: ${Math.round(endAll - startAll)}ms`)
  })
}, [])
```

All requests are performed in parallel (async) with a staging user. I chose  judy.morrison@id.me and Karsyn as my two staging users because they are authorized for many different services.

After logging in with a staging user in the simulator, I reloaded the app repeatedly, checking the log output to get request timing data. I checked the requests in Charles to ensure valid data was being returned. I also used Charles to throttle the network speed, simulating endpoint load times over slower cell connections.

## Results

With the judy.morrison@id.me staging user, all the above requests completed. Here's an example run:

```
/v0/claims-and-appeals-overview: 524ms
/v0/appointments: 541ms
/v0/facilities-info: 718ms
/v0/claims/decision-letters: 923ms
/v0/disability-rating: 938ms
/v0/maintenance_windows: 1076ms
/v0/health/rx/prescriptions: 1271ms
/v0/messaging/health/folders: 1365ms
/v0/messaging/health/recipients: 1413ms
/v0/messaging/health/folders/0/messages: 1464ms
/v0/letters: 1782ms
/v0/military-service-history: 1838ms
/v0/letters/beneficiary: 2008ms
/v0/user/gender_identity/edit: 2143ms
/v1/health/immunizations: 2285ms
/v0/payment-history: 2440ms
/v0/payment-information/benefits: 2459ms
/v0/user/authorized-services: 2746ms
/v0/user/demographics: 4342ms
/v1/user: 5473ms

All requests finished: 5476ms
```

Average of 10 loads at 6pm Eastern with judy.morrison@id.me: (5407 + 5541 + 5871 + 5822 + 5754 + 7170 + 5967 + 5088 + 5264 + 6071) / 10 = 5796ms
Average of 10 loads at 11am Eastern with judy.morrison@id.me: (5521 + 6431 + 5435 + 5642 + 5844 + 5674 + 5630 + 5807 + 5434 + 6235) / 10 = 5765ms
For comparison, loading only the `/v1/user` request by itself took 3884ms with judy.morrison@id.me.

The Karsyn user on staging has a variety of services enabled, but some are disabled. I commented out the requests corresponding to the disabled services. Here's what remained: 

```
/v0/appointments: 299ms
/v0/messaging/health/folders: 343ms
/v0/maintenance_windows: 346ms
/v0/facilities-info: 381ms
/v0/messaging/health/recipients: 423ms
/v0/user/authorized-services: 455ms
/v0/user/gender_identity/edit: 519ms
/v0/messaging/health/folders/0/messages: 556ms
/v1/health/immunizations: 584ms
/v0/military-service-history: 748ms
/v0/health/rx/prescriptions: 1007ms
/v0/user/demographics: 1602ms
/v1/user: 1901ms

All requests finished: 1903ms
```

Average of 10 loads at 6pm Eastern with Karsyn: (2183 + 1759 + 1706 + 1751 + 1745 + 1823 + 1759 + 1749 + 1734 + 2157) / 10 = 1837ms
Average of 10 loads at 11am Eastern with Karsyn: (2133 + 1762 + 1930 + 1842 + 1811 + 1782 + 1788 + 1801 + 1914 + 2057) / 10 = 1882ms
For comparison, loading only the `/v1/user` request by itself took 1653ms with Karsyn.

I also tried loading judy.morrison@id.me with Charles simulating a 3G cellular connection (4mbps down/1mpbs up):

```
/v0/claims-and-appeals-overview: 253ms
/v0/appointments: 681ms
/v0/disability-rating: 818ms
/v0/claims/decision-letters: 1057ms
/v0/letters: 1364ms
/v0/facilities-info: 1666ms
/v0/letters/beneficiary: 1881ms
/v0/maintenance_windows: 1894ms
/v0/payment-history: 1962ms
/v0/payment-information/benefits: 2118ms
/v0/messaging/health/recipients: 2177ms
/v0/messaging/health/folders: 2305ms
/v0/health/rx/prescriptions: 2329ms
/v0/messaging/health/folders/0/messages: 2337ms
/v0/user/gender_identity/edit: 2568ms
/v1/health/immunizations: 2582ms
/v0/military-service-history: 2747ms
/v0/user/authorized-services: 3212ms
/v0/user/demographics: 5036ms
/v1/user: 5982ms

All requests finished: 6003ms
```

These times are only slightly slower than Judy's times with a full-speed network connection.

I tried loading judy.morrison@id.me with Charles simulating a very slow cellular connection (200kbps down/100kbps up):

```
/v0/claims-and-appeals-overview: 1169ms
/v0/disability-rating: 1295ms
/v0/claims/decision-letters: 1473ms
/v0/letters: 1900ms
/v0/appointments: 1906ms
/v0/facilities-info: 2316ms
/v0/payment-information/benefits: 2378ms
/v0/letters/beneficiary: 2458ms
/v0/payment-history: 2545ms
/v0/maintenance_windows: 2666ms
/v0/messaging/health/recipients: 2905ms
/v0/messaging/health/folders: 2934ms
/v0/messaging/health/folders/0/messages: 3039ms
/v0/health/rx/prescriptions: 3162ms
/v0/military-service-history: 3392ms
/v0/user/gender_identity/edit: 3397ms
/v1/health/immunizations: 3567ms
/v0/user/authorized-services: 3931ms
/v0/user/demographics: 5643ms
/v1/user: 6978ms

All requests finished: 6981ms
```

These times are a bit slower but still not bad considering the extreme slowness of the network connection.


### Summary

- Multiple parallel requests load without issue.
- Loading the full list of endpoints is slower than loading just a single /v1/user request, but not hugely slower.
- Network connection speed doesn't affect request times too much, indicating that processing the request takes most of the request time. Data transfer time is minimal.
- The `/v1/user` endpoint is consistently the slowest, followed by `/v0/user/demographics`. The `/v1/user` endpoint will soon be replaced by a v2 version which should be faster.
- Loading the data for judy.morrison is consistently slower than for Karsyn. This is likely because we can load the full set of endpoints for Judy, whereas Karsyn has some services disabled.

Note that load times on staging may not directly correlate with load times in production because of various differences such as server setup, network setup, and server/network utilization. BE engineers could provide more detail on these differences. Since we don't have access to any production users, it's not possible to perform these benchmarks in production.

### Future improvements

It would be useful to gather performance metrics on production endpoints. We could potentially use analytics events to do so. Such metrics would help us understand the relative performance of the various endpoints, as well as how performance changes over time. @timwright12 will create a ticket to explore this further.

**Comment by alexandec** [comment]
Moved this ticket to Code Review just to indicate that I've completed the initial work. I'll leave the ticket open for a bit in case there are comments or questions.

**Comment by alexandec** [comment]
Some other options that come to mind are [Firebase Performance Monitoring](https://firebase.google.com/docs/perf-mon) or Datadog.

**Comment by dumathane** [comment]
@alexandec It's worth pointing out that all of the above metrics are still using v1 user, so we are essentially calling to retrieve the same data twice? We are using a lot of the new endpoints for pieces of v1 and still just calling v1 in it's entirety. Once we merge the v2 change to develop do you think we should rerun your metrics on your branch and track a difference? Seems easy enough.

As for performance monitoring, we attempted to do this previously with some backend apis I believe with a start/stop timer we could send to analytics. Chris are you suggesting basically revamping this to you know, actually work? 😄

**Comment by alexandec** [comment]
@dumathane sure, we can re-run these once we're using v2 instead. Should we keep this ticket open or create a new one? I'm thinking a new one to cover the additional effort, but curious what you think.

I've seen a start-stop timer that's supposed to track how long a user is on a given page (which doesn't work right anymore), but I don't recall seeing what we've tried for API calls. But yes, something that actually works, haha.

**Comment by alexandec** [comment]
Here's the followup ticket with times for the v2 user endpoint: #6952. I performed benchmarks and added my findings there.
